File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/variables/ProjectVariables.java
package in.org.quicko.silo.client.variables;

import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.Transport;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * The declarations within one project: {@code declare}, {@code rename},
 * {@code describe}, {@code undeclare}. Values live on
 * {@link EnvironmentVariables} instead, because a name is declared once and
 * valued per environment.
 */
public final class ProjectVariables {
  private final Transport transport;
  private final String project;

  public ProjectVariables(Transport transport, String project) {
    this.transport = transport;
    this.project = project;
  }

  public Variable declare(String name) {
    return declare(name, DeclareVariableOptions.none());
  }

  public Variable declare(String name, DeclareVariableOptions options) {
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("name", name);
    if (options.description() != null) body.put("description", options.description());
    if (options.value() != null) body.put("value", options.value());

    return Variable.fromWire(transport.json(
        TransportRequest.post(ApiPath.projectVariables(project))
            .query("env", options.environment())
            .body(body)
            .options(options.request())
            .build()));
  }

  public Variable rename(String name, String newName) {
    return rename(name, newName, VariableEnvironmentOptions.none());
  }

  public Variable rename(String name, String newName, VariableEnvironmentOptions options) {
    return updateDeclaration(name, Map.of("name", newName), options);
  }

  public Variable describe(String name, String description) {
    return describe(name, description, VariableEnvironmentOptions.none());
  }

  public Variable describe(String name, String description, VariableEnvironmentOptions options) {
    return updateDeclaration(name, Map.of("description", description), options);
  }

  public void undeclare(String name) {
    undeclare(name, VariableEnvironmentOptions.none());
  }

  /** Forgets the name, and every environment's value with it. */
  public void undeclare(String name, VariableEnvironmentOptions options) {
    transport.empty(
        TransportRequest.delete(ApiPath.projectVariable(project, name))
            .query("env", options.environment())
            .options(options.request())
            .build());
  }

  private Variable updateDeclaration(
      String name, Map<String, Object> body, VariableEnvironmentOptions options) {
    return Variable.fromWire(transport.json(
        TransportRequest.patch(ApiPath.projectVariable(project, name))
            .query("env", options.environment())
            .body(body)
            .options(options.request())
            .build()));
  }
}

File: packages/silo-client-dart/lib/src/variables/project_variables.dart
import '../transport/api_path.dart';
import '../transport/transport.dart';
import '../transport/transport_request.dart';
import 'declare_variable_options.dart';
import 'variable.dart';
import 'variable_environment_options.dart';

/// Declarations, project-wide. Values live on `EnvironmentVariables`.
final class ProjectVariables {
  const ProjectVariables(this._transport, this._project);

  final Transport _transport;
  final String _project;

  Future<Variable> declare(String name, [DeclareVariableOptions options = const DeclareVariableOptions()]) async =>
      Variable.fromJson(
        await _transport.json(
          TransportRequest(
            'POST',
            ApiPath.projectVariables(_project),
            query: {'env': ?options.environment},
            body: {'name': name, 'description': ?options.description, 'value': ?options.value},
            options: options,
          ),
        ),
      );

  Future<Variable> rename(
    String name,
    String to, [
    VariableEnvironmentOptions options = const VariableEnvironmentOptions(),
  ]) => _update(name, {'name': to}, options);

  Future<Variable> describe(
    String name,
    String description, [
    VariableEnvironmentOptions options = const VariableEnvironmentOptions(),
  ]) => _update(name, {'description': description}, options);

  /// Forgets the name and every environment's value.
  Future<void> undeclare(String name, [VariableEnvironmentOptions options = const VariableEnvironmentOptions()]) =>
      _transport.empty(
        TransportRequest(
          'DELETE',
          ApiPath.projectVariable(_project, name),
          query: {'env': ?options.environment},
          options: options,
        ),
      );

  Future<Variable> _update(String name, Map<String, Object?> body, VariableEnvironmentOptions options) async =>
      Variable.fromJson(
        await _transport.json(
          TransportRequest(
            'PATCH',
            ApiPath.projectVariable(_project, name),
            query: {'env': ?options.environment},
            body: body,
            options: options,
          ),
        ),
      );
}
